### PR TITLE
storage-client: internal command to update ingestions

### PR DESCRIPTION
**Pull request by sploiselle**
As part of `ALTER SOURCE`, we will need the ability to redescribe and restart an ingestion. Note that this is just prep work for making this accessible later, so nothing to test and we will make sure these code paths are thoroughly exercised once they're connected to the adapter.

### Motivation

This PR adds a known-desirable feature. Part of #15832

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/storage-client/src/client.rs
Comment (aljoscha) [inline src/storage-client/src/client.rs:114]: copy pasta 🍝

File: src/storage-client/src/client.rs
Comment (aljoscha) [inline src/storage-client/src/client.rs]: Could it be that we have to remove uppers for exports that now don't exist anymore? Same question for the rehydration shim.

File: src/storage/src/storage_state.rs
Comment (aljoscha) [inline src/storage/src/storage_state.rs:1168]: typo: "for so"

File: src/storage-client/src/client.rs
Comment (aljoscha) [inline src/storage-client/src/client.rs:95]: You've been bitten by this?

File: src/storage-client/src/controller.rs
Comment (aljoscha) [inline src/storage-client/src/controller.rs:3034]: just musing: maybe `populate_ingestion`, or `enrich_ingestion`? But you're better with words than I am... 😅

File: src/storage-client/src/controller.rs
Comment (aljoscha) [inline src/storage-client/src/controller.rs:2947]: I believe we talked about this in the past and I'm still fuzzy on the language here: in my head, we're installing, for each element of `collections`, a read hold on all of the `storage_dependencies`.

Let's chat in person to resolve this quickly! 😅 But it's always tricky for me to get a common understanding of the dependency "arrow".

Comment (aljoscha) [reply src/storage-client/src/controller.rs:2947]: Also, excellent move, factoring all these out into methods! 🥇

File: src/storage-client/src/controller.rs
Comment (aljoscha) [inline src/storage-client/src/controller.rs:161]: 😅

File: src/storage-client/src/types/sources.rs
Comment (aljoscha) [inline src/storage-client/src/types/sources.rs:133]: I have a feeling `e, s, o` will not age well

File: src/adapter/src/util.rs
Comment (aljoscha) [inline src/adapter/src/util.rs:324]: structured errors! ✊

File: src/storage-client/src/controller.rs
Comment (aljoscha) [inline src/storage-client/src/controller.rs:861]: who should do what with this `halt`? I can look at the code, but I'm also lazy and rustdoc would help... 😅

**Review by aljoscha — COMMENTED** [review]
excellent! I had some comments/questions inline and I'll let Petros do the real review

File: src/storage-client/src/client.rs
Comment (sploiselle) [reply src/storage-client/src/client.rs]: There's an impedance mismatch between creating and dropping subsources where the lower-level clients don't get uppers for subsources until their primary source/ingestion gets created, but they get dropped as if they were primary sources, i.e. the coordinator is responsible for dropping them, where the `AllowCompaction` machinery removes them from the lower-level clients.

File: src/storage-client/src/controller.rs
Comment (sploiselle) [reply src/storage-client/src/controller.rs:2947]: Ah, yes, I recall now that we had different understandings of this and it took me a bit of thinking but ended up seeing that you were correct. I'm glad to adjust the language here but also glad to chat.

File: src/storage-client/src/controller.rs
Comment (sploiselle) [reply src/storage-client/src/controller.rs]: enrich it is

File: src/storage-client/src/client.rs
Comment (aljoscha) [reply src/storage-client/src/client.rs]: let's chat about this before you merge. I'm not sure yet that I understand completely 😅

File: src/storage-client/src/client.rs
Comment (petrosagg) [inline src/storage-client/src/client.rs:636]: We have a specific error for this, `StorageError::SourceIdReused(id)`. Can you return a Result with that and assert it in the caller?

File: src/storage/src/storage_state.rs
Comment (petrosagg) [inline src/storage/src/storage_state.rs:1095]: This doesn't look right because reconciliation might happen after the controller sends alter commands and so the rehydration client might send newer state here.

File: src/storage-client/src/client.rs
Comment (sploiselle) [reply src/storage-client/src/client.rs:95]: If you look at the reconciliation behavior:
https://github.com/MaterializeInc/materialize/blob/main/src/storage/src/storage_state.rs#L1003

It iterates over the set and only retains some of the operations, the others are "reconciled" away.

File: src/storage-client/src/client.rs
Comment (sploiselle) [reply src/storage-client/src/client.rs]: We chatted about this over zoom and came to a shared understanding of what's going on here.

File: src/storage-client/src/controller.rs
Comment (sploiselle) [reply src/storage-client/src/controller.rs]: Removed the interior halt from the error and added an explicit halt in `alter_collection`.

File: src/storage-client/src/controller.rs
Comment (sploiselle) [reply src/storage-client/src/controller.rs:319]: There are still other errors that might occur, e.g. `install_dependency_read_holds`.

File: src/storage-client/src/types/sources.rs
Comment (sploiselle) [reply src/storage-client/src/types/sources.rs:144]: Refactored this as we discussed to produce worse messages, but be simpler to reason about.

As part of this, also implemented a `merge_join` for `BTreeMap`, which is something I want from time to time and is useful here.

File: src/storage/src/storage_state.rs
Comment (sploiselle) [reply src/storage/src/storage_state.rs:1239]: Chatted about this and it was part of the oversight that ingestions can add new subsources

File: src/storage/src/storage_state.rs
Comment (sploiselle) [reply src/storage/src/storage_state.rs:1023]: Think this is done, but please check my work!

File: src/ore/src/btree_map.rs
Comment (benesch) [inline src/ore/src/btree_map.rs:44]: Did you consider [`itertools::merge_join_by`](https://docs.rs/itertools/latest/itertools/fn.merge_join_by.html)?

File: src/storage/src/storage_state.rs
Comment (sploiselle) [reply src/storage/src/storage_state.rs]: Believe I addressed this but please check my work

File: src/storage-client/src/types/sources.rs
Comment (sploiselle) [reply src/storage-client/src/types/sources.rs:150]: Refactored this again to just use an itertools function I wasn't aware of previously.

File: src/storage-client/src/client.rs
Comment (sploiselle) [reply src/storage-client/src/client.rs]: nice ty

File: src/storage-client/src/client.rs
Comment (sploiselle) [reply src/storage-client/src/client.rs]: Using the error was annoying here so I put the `GlobalId` in the error and just panic on it. lmk if you had some other vision that's more ergonomic.

File: src/storage-client/src/client.rs
Comment (sploiselle) [reply src/storage-client/src/client.rs]: Oh sure

File: src/ore/Cargo.toml
Comment (benesch) [inline src/ore/Cargo.toml:66]: Vestigial from past iteration of the PR, I imagine!

Comment (sploiselle) [reply src/ore/Cargo.toml:66]: Oh yes, will make sure this gets reverted

**Comment by sploiselle** [comment]
@petrosagg Any word on this review?

File: src/storage-client/src/types/sources.rs
Comment (petrosagg) [inline src/storage-client/src/types/sources.rs:141]: nit: whenever this happens (multi-line for declaration) you can store the iteratee in a variable and use that in the for loop which results in the same  lines of code and a clearer for definition:

```rust

        let compatibility_checks = [
            source_imports == &other.source_imports,
            ingestion_metadata == &other.ingestion_metadata,
            source_exports
                .iter()
                .merge_join_by(&other.source_exports, |(l_key, _), (r_key, _)| {
                    l_key.cmp(r_key)
                })
                .all(|r| match r {
                    Both((_, l_val), (_, r_val)) => l_val == r_val,
                    _ => true,
                }),
            instance_id == &other.instance_id,
            remap_collection_id == &other.remap_collection_id,
        ];
        for incompatible in compatibility_checks {
```

File: src/storage/src/storage_state.rs
Comment (sploiselle) [reply src/storage/src/storage_state.rs]: This was my misunderstanding from a convo with Aljoscha. The new code handles both alters and creates identically, so this is a non-issue.

File: src/storage/src/storage_state.rs
Comment (sploiselle) [reply src/storage/src/storage_state.rs]: Yeah, this was my oversight. In a conversation, you had mentioned we just needed to call `SuspendAndRestart` to get these working––however, this needs to run code analogous to creating a source to handle the case where the ingestion has not been updated on the worker.

File: src/storage-client/src/controller/rehydration.rs
Comment (sploiselle) [reply src/storage-client/src/controller/rehydration.rs]: Aljoscha had the same question: https://github.com/MaterializeInc/materialize/pull/19310#discussion_r1196182899

There's an impedance mismatch between creating and dropping subsources; creating a subsource is a nop from the perspective of the lower-level clients. Their frontiers, etc. are only tracked once the primary source gets created. However, when you drop a subsource, it has its own `AllowCompaction` commands that propagate through the clients.

**Comment by sploiselle** [comment]
@petrosagg You had asked, "What is the difference between a CreateSource and an AlterSource" and it turns out that there is no meaningful difference aside from the style of adjustments permitted to internal state (e.g. CreateSource must be inserting a ingestion and AlterSource must be updating an existing ingestion).

Because of this, I was able to consolidate both of these into a single command with a boolean toggle that encapsulates the allowed state changes (I called it `RunIngestions` but am not attached to the name).

Also, just wanted to touch base about the concern of the ordering of ingestion states that we talked about yesterday and say that the code is constructed to swallow responses for IDs that have not been seen in the `StorageState` yet.

File: src/storage/src/storage_state.rs
Comment (petrosagg) [inline src/storage/src/storage_state.rs:1021]: Same here, that seems like an error worth panicking for and reporting to sentry

**Review by petrosagg — APPROVED** [review]
Looks good modulo figuring out those `halt!` usages

**Pull request merged**
